Rubberduck.Parsing/VBA/COMReferenceManagerBase.cs

```diff
@@ -0,0 +1,269 @@
+        private readonly string _serializedDeclarationsPath;
+
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+
+        public COMReferenceManagerBase(RubberduckParserState state, IParserStateManager parserStateManager, string serializedDeclarationsPath = null)
+        {
+            if (state == null) throw new ArgumentNullException(nameof(state));
+            if (parserStateManager == null) throw new ArgumentNullException(nameof(parserStateManager));
+
+            _state = state;
+            _parserStateManager = parserStateManager;
+            _serializedDeclarationsPath = serializedDeclarationsPath
+                ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Rubberduck", "declarations");
+        }
+
+
+        private bool _lastRunLoadedReferences;
+        public bool LastRunLoadedReferences
+        {
+            get
+            {
+                return _lastRunLoadedReferences;
+            }
+        }
+
+        private bool _lastRunUnloadedReferences;
+        public bool LastRunUnloadedReferences
+        {
+            get
+            {
+                return _lastRunUnloadedReferences;
+            }
+        }
+
+        private readonly HashSet<ReferencePriorityMap> _projectReferences = new HashSet<ReferencePriorityMap>();
+        public ICollection<ReferencePriorityMap> ProjectReferences
+        {
+            get
+            {
+                return _projectReferences.ToHashSet();
+            }
+        }
+
+
+        protected abstract void LoadReferences(List<IReference> referencesToLoad, ConcurrentBag<IReference> unmapped, CancellationToken token);
+
+
+        public void SyncComReferences(IReadOnlyList<IVBProject> projects, CancellationToken token)
+        {
+            _lastRunLoadedReferences = false;
+            _lastRunUnloadedReferences = false;
+
+            var unmapped = new ConcurrentBag<IReference>();
+
+            var referencesToLoad = GetReferencesToLoadAndSaveReferencePriority(projects);
+
+            _state.OnStatusMessageUpdate(ParserState.LoadingReference.ToString()); //Why do we use this method to change the displayed state?
+
+            if (referencesToLoad.Any())
+            {
+                _lastRunLoadedReferences = true;
+                LoadReferences(referencesToLoad, unmapped, token);
+            }
+
+            var notMappedReferences = NonMappedReferences(projects);
+            foreach (var item in notMappedReferences)
+            {
+                unmapped.Add(item);
+            }
+
+            if (unmapped.Any())
+            {
+                _lastRunUnloadedReferences = true;
+                foreach (var reference in unmapped)
+                {
+                    UnloadComReference(reference, projects);
+                }
+            }
+        }
+
+        private List<IReference> GetReferencesToLoadAndSaveReferencePriority(IReadOnlyList<IVBProject> projects)
```

Comment:
materializing here doesn't give us any benefits. it should be easy to refactor this into an `IEnumerable` and use `yield`-returns